src/main/java/formflow/library/FormFlowController.java

```diff
@@ -46,9 +61,110 @@ protected Boolean doesFlowExist(String flow) {
     );
   }
 
-  protected static void throwNotFoundError(String flow,String screen, String message) {
-    throw new ResponseStatusException(HttpStatus.NOT_FOUND, String.format("There was a problem with the request (flow: %s, screen: %s): %s",
-        flow, screen, message));
+  protected static void throwNotFoundError(String flow, String screen, String message) {
+    throw new ResponseStatusException(HttpStatus.NOT_FOUND,
+        String.format("There was a problem with the request (flow: %s, screen: %s): %s",
+            flow, screen, message));
+
+  }
+
+  /**
+   * If the submission information exists in the HttpSession, find it in the db. If a submission is not found, create a new one.
+   *
+   * @param httpSession The HttpSession to look in for information
+   * @param flow        The name of the flow to retrieve data about
+   * @return Submission A Submission object from the database or a new one if one was not found
+   */
+  public Submission findOrCreateSubmission(HttpSession httpSession, String flow) {
+    Submission submission = null;
+    try {
+      submission = getSubmissionFromSession(httpSession, flow);
+    } catch (ResponseStatusException ignored) {
+      // it's okay if it doesn't exist already
+    }
+
+    if (submission == null) {
+      log.info("Submission not found in session for flow '{}', creating one.", flow);
+      submission = new Submission();
+    }
+    return submission;
+  }
+
+  /**
+   * Returns the UUID of the Submission associated with the given flow.
+   *
+   * @param session The HttpSession the user is in
+   * @param flow    The flow to look up the submission ID for
+   * @return The submission id if it exists for the given flow, else null
+   */
+  public static UUID getSubmissionIdForFlow(HttpSession session, String flow) {
```

Comment:
How does this abstraction look? I liked putting this up here in case we change how we store the submissions again. If we do, then we will have less spots to change the logic of.

Reply:
There are three methods I moved to being here `getSubmissionIdForFlow`, `setSubmissionIdInSession`, and `findOrCreateSubmission`.

Reply:
Feels good to me!